sync-android/src/main/java/com/cloudant/sync/datastore/encryption/KeyStorage.java

```diff
@@ -0,0 +1,186 @@
+     * Notice that if there is already data in the {@link SharedPreferences} bound to the same
+     * identifier used to create
+     * this storage, the operation will fail.
+     *
+     * @param data {@link KeyData} to save to the {@link SharedPreferences}
+     * @return true (success) or false (fail)
+     */
+    public boolean saveEncryptionKeyData(KeyData data) {
+        JSONObject objectToSave = new JSONObject();
+        try {
+            objectToSave.put(KEY_DPK, DKPEncryptionUtil.byteArrayToHexString(data
+                    .getEncryptedDPK()));
+            objectToSave.put(KEY_IV, DKPEncryptionUtil.byteArrayToHexString(data.getIv()));
+            objectToSave.put(KEY_SALT, DKPEncryptionUtil.byteArrayToHexString(data.getSalt()));
+            objectToSave.put(KEY_ITERATIONS, data.getIterations());
+            objectToSave.put(KEY_VERSION, data.getVersion());
+
+            String valueToSave = objectToSave.toString();
+
+            SharedPreferences.Editor editor = this.prefs.edit();
+            editor.putString(buildKey(), valueToSave);
+            editor.commit();
+
+        } catch (JSONException e) {
+            LOGGER.log(Level.SEVERE, e.getLocalizedMessage(), e);
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Remove from the {@link SharedPreferences} a {@link KeyData} associated to the same
+     * identifier used to
+     * create this storage.
+     * <p/>
+     * It will succeed if the data is deleted or if there is no data at all.
+     *
+     * @return true (success) or false (fail)
+     */
+    public boolean clearEncryptionKeyData() {
+        try {
+            SharedPreferences.Editor editor = this.prefs.edit();
+            editor.remove(buildKey());
+            editor.commit();
+        } catch (Exception e) {
+            LOGGER.log(Level.SEVERE, e.getLocalizedMessage(), e);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Look for data saved in the {@link SharedPreferences} with the same identifier used to
+     * create this storage instance.
+     *
+     * @return true (data found) or false (data not found)
+     */
+    public boolean encryptionKeyDataExists() {
+        return this.prefs.contains(buildKey());
+    }
+
+    private String buildKey() {
+        return PREF_NAME_DPK + "-" + this.account;
+    }
+
+    private JSONObject convertStringToJSON(String jsonString) throws JSONException {
+        int beginIndex = jsonString.indexOf("{");
+        int endIndex = jsonString.lastIndexOf("}");
+
+        if (beginIndex == -1 || endIndex == -1 || beginIndex > endIndex + 1) {
+            String message = "Input string does not contain brackets, or input string is invalid." +
+                    " The string is: " + jsonString;
+            //logger.debug (message);
+            throw new JSONException(message);
+        }
+
+        String secureJSONString = jsonString.substring(beginIndex, endIndex + 1);
```

Comment:
I'm not convinced the above dance to find the brackets offers any security? Is it worthwhile?